File: LP2stuff/BFS_DFS.cpp
#include <bits/stdc++.h>
#include <iostream>

using namespace std;
class Queue{
public:
    int size1;
    int f;        //front for deletion
    int r;        //rear for insertion
    int *a;
};
void enqueue(Queue *st,int data){
    if(st->r==st->size1-1){
        cout<<"Queue FULL"<<endl;
    }
    else{
        st->r++;
        st->a[st->r]=data;
    }
}
int dequeue(Queue *st){
    int x=-1;
    if(st->f==st->r){
        cout<<"Queue Empty"<<endl;
    }
    else{
        st->f++;
        x=st->a[st->f];
    }
    return x;
}
int isempty(Queue *st){
    if(st->f==st->r){
        return 1;
    }
    return 0;
}
int arr[8][8]= {{0,0,0,0,0,0,0,0},
                {0,0,1,1,1,0,0,0},
                {0,1,0,1,0,0,0,0},
                {0,1,1,0,1,1,0,0},
                {0,1,0,1,0,1,0,0},
                {0,0,0,1,1,0,1,1},
                {0,0,0,0,0,1,0,0},
                {0,0,0,0,0,1,0,0}};
class graph{
public:
    int *visited;
};
void bfs(graph *t,Queue *q,int i){
    cout<<i<<" ";
    t->visited[i]=1;
    enqueue(q,i);
    while(!isempty(q)){
        int u=dequeue(q);
        for(int j=1;j<=8;j++){
            if(arr[u][j]==1&&t->visited[j]==0){
                cout<<j<<" ";
                t->visited[j]=1;
                enqueue(q,j);
            }
        }
    }

}
void dfs(int u){
    static int visited[8]={0};
    if(visited[u]==0){
        cout<<u<<" ";
        visited[u]=1;
        for(int i=1;i<=8;i++){
            if(arr[u][i]==1 && visited[i]==0){
                dfs(i);
            }
        }
    }
}
int main()
{
    cout<<"*******MENU*******"<<'\n';
    cout<<"1.BFS"<<'\n';
    cout<<"2.DFS"<<'\n';
    cout<<"3.Exit"<<'\n';
    int e=0;
    int c;
    while(e==0){
        cout<<'\n'<<"Enter Choice"<<'\n';
        cin>>c;
        switch(c){
            case 1:{
                graph t;
                Queue s;
                s.f=-1;
                s.r=-1;
                s.size1=8;
                s.a=new int[8];
                t.visited=new int[8];
                for(int i=0;i<8;i++){
                    t.visited[i]=0;
                }
                int c1;
                cout<<"Enter starting Vertex(1-7)"<<'\n';
                cin>>c1;
                bfs(&t,&s,c1);
                break;
            }
            case 2:{
                int c1;
                cout<<"Enter starting Vertex(1-7)"<<'\n';
                cin>>c1;
                dfs(c1);
                break;
            }
            case 3:{
                cout<<"Thank-You"<<'\n';
                e=1;
                break;
            }
        }
    }
    return 0;
}

File: LP2stuff/bfs_dfs.py
Assignment 1
Implement Depth First Search algorithm and Breadth First Search algorithm. Use an undirected graph
and develop a recursive algorithm for searching all the vertices of a graph or tree data structure.
'''

class Node:
    def __init__(self, i):
        self.l = None
        self.r = None
        self.i = i

# -------------------------------------------------------
    
class Tree:
    def __init__(self, rX):
        self.root = rX

    # Get input function
    def getInput(self):
        q = [self.root]
        print("\nEnter data or -1 for no data\n")
        while q:
            tmp = q.pop()
            if(tmp!=None):
                lC = int(input("Enter left child of {}: ".format(tmp.i)))
                if lC!=-1:
                    tmp.l = Node(lC)
                    q.append(tmp.l)
                else:
                    tmp.l=None

                rC = int(input("Enter right child of {}: ".format(tmp.i)))
                if rC!=-1:
                    tmp.r = Node(rC)
                    q.append(tmp.r)
                else:
                    tmp.r=None

    # Recursive DFS function
    def dfs(self, x):
        print(x.i, end = ' ')
        if x.l:
            self.dfs(x.l)
        if x.r:
            self.dfs(x.r)

    # Recursive BFS function
    def bfs(self, x):
        queue = []
        queue.append(x)

        while(len(queue)>0):
            print(queue[0].i, end=' ')
            tmp = queue.pop(0)

            if tmp.l is not None:
                queue.append(tmp.l)
            if tmp.r is not None:
                queue.append(tmp.r)
    
# -------------------------------------------------------

# Menu
while(True):
    print("\n------------------------")
    print("BFS and DFS traversal")
    print("1. Enter data")
    print("2. BFS traversal")
    print("3. DFS traversal")
    print("4. Exit\n")

    c = int(input("Enter choice: "))
    if c==1:
        # Create a tree
        r = int(input("Enter the root node: "))
        ROOT = Node(r)
        t = Tree(ROOT)
        t.getInput()

    elif c==2:
        print("\n-- BFS Traversal --")
        t.bfs(ROOT)
        print("")

    elif c==3:
        print("\n-- DFS Traversal --")
        t.dfs(ROOT)
        print("")

    elif c==4:
        break

    else:
        print("Invalid Input")

'''
          20
        /    \
       7     23 
      / \    / \
     3   8  15  27

    BFS:    20 7 23 3 8 15 27
    DFS:    20 7 3 8 23 15 27

          10
        /    \
       2     12
      / \
     1   3

    BFS:    10 2 12 1 3
    DFS:    10 2 1 3 12
'''
